Refactor remote federation / fallback

gabrielhartmann opened the pull request:
### Commit message
This simplifies introduction of remote federation, avoiding
the necessity of custom modification of every method in the
AggregatingJobServiceGateway.

### Description of the Change
The motivation for this change was the start of work for adding UpdateJobCapacity to fedv2.  It became clear that modifying the behavior of each API call in the AggregatingJobServiceGateway was going to get messy.  The approach presented here seems cleaner to me.

The FallBackJobServiceGateway has a primary and a secondary: 
```
primary:   RemoteJobServiceGateway      (aka fedv2)
secondary: AggregatingJobServiceGateway (aka fedv1)
```

If remote federation is enabled then the primary is attempted for all JobServiceGateway calls.  If an error occurs, fallback to the secondary occurs.  If remote federation is disabled the FallbackJobServiceGateway does not attempt the primary and calls the secondary in all cases.  So when remote federation is enabled the FallbackJobServiceGateway attempts to use fedv2 and falls back to fedv1 when an error is encountered.  When remote federation is disabled all calls are sent directly to fedv1.

Metrics look like this on fallback depending on GRPC status codes:
```
titus.fallbackJobServiceGateway.fallbackCount.createJob UNAVAILABLE
titus.fallbackJobServiceGateway.fallbackCount.findJobs UNIMPLEMENTED
```

gabrielhartmann commented:
I've tested this all on my dev stack.  I'll look into the unit test failures and fix the dependency injection stuff its complaining about.

fabiokung commented on titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java:
```diff
@@ -1,64 +1,254 @@
-    private final JobManagementServiceGrpc.JobManagementServiceStub primaryClient;
-    private final GrpcConfiguration grpcConfiguration;
+public class FallbackJobServiceGateway implements JobServiceGateway {
+    public static String METRIC_ROOT = "titus.fallbackJobServiceGateway.";
+    private static String METRIC_FALLBACK_ROOT= METRIC_ROOT + "fallbackCount.";
```
nit: constants are `final`

fabiokung commented on titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java:
```diff
@@ -1,64 +1,254 @@
+    private Id getMetricId(String metricName, Throwable t) {
+        String reason = t.getClass().getSimpleName();
+
+        if (t instanceof StatusRuntimeException) {
```
I recommend using `Status.fromThrowable(t).getCode()` instead

fabiokung commented on titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java:
```diff
@@ -1,64 +1,254 @@
+        return getFallbackObservable(methodName, primaryObservable, secondaryObservable);
+    }
+
+    @Override
+    public Observable<JobChangeNotification> observeJob(String jobId, CallMetadata callMetadata) {
+        final String methodName = "observeJob";
+        Observable<JobChangeNotification> primaryObservable = primary.observeJob(jobId, callMetadata);
+        Observable<JobChangeNotification> secondaryObservable = secondary.observeJob(jobId, callMetadata);
+        return getFallbackObservable(methodName, primaryObservable, secondaryObservable);
+    }
+
+    @Override
+    public Observable<JobChangeNotification> observeJobs(ObserveJobsQuery query, CallMetadata callMetadata) {
+        final String methodName = "observeJobs";
+        Observable<JobChangeNotification> primaryObservable = primary.observeJobs(query, callMetadata);
+        Observable<JobChangeNotification> secondaryObservable = secondary.observeJobs(query, callMetadata);
+        return getFallbackObservable(methodName, primaryObservable, secondaryObservable);
+    }
```
This strategy will not work well with stream APIs. If there is an error after the stream is established and some items were already sent, we want to propagate the error, not redo the stream on the secondaryObservables

fabiokung commented on titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java:
```diff
@@ -1,64 +1,254 @@
+
+    private <T> Observable<T> getFallbackObservable(String methodName, Observable<T> primary, Observable<T> secondary) {
+        if (federationConfiguration.isRemoteFederationEnabled()) {
+            return primary.onErrorResumeNext( (t) -> {
```
For operations with side effects, we run the risk of retrying un-retryable errors here, i.e. on partial failures where the remote service partially succeeded but the call returns an error - e.g. a timeout.

This is OK for a first iteration, but did you consider limiting what types of errors cause a fallback to `secondary`? Perhaps only `UNIMPLEMENTED` should trigger a fallback?

gabrielhartmann replied:
This is a good idea `UNIMPLEMENTED` and `UNAVAILABLE` would probably be good choices.  Let me think about it a little.

fabiokung replied:
sgtm, I'd caution against `UNAVAILABLE` as well, I'm not sure it could be thrown for in-flight requests. Perhaps I'm being a bit too paranoid.

fabiokung reviewed (changes requested):
LGTM overall, but I'd leave the streaming APIs alone, or you need some more smartness there to know to fallback only when the `primary` stream wasn't estabilished

gabrielhartmann commented:
@fabiokung Thanks for the review.  I believe I've addressed all your comments.  Are there other streaming APIs apart from `observeJob` and `observeJobs` which I should also avoid falling back on?

fabiokung commented on titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java:
```diff
@@ -1,64 +1,272 @@
+
+    @Override
+    public Observable<JobChangeNotification> observeJob(String jobId, CallMetadata callMetadata) {
+        // TODO: Direct stream APIs to the remote (primary) JobServiceGateway.
```
fix the comment here? We are sending to secondary

gabrielhartmann replied:
The TODO is to direct to the primary instead of the secondary.

fabiokung commented on titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java:
```diff
@@ -1,64 +1,272 @@
+
+    @Override
+    public Observable<JobChangeNotification> observeJobs(ObserveJobsQuery query, CallMetadata callMetadata) {
+        // TODO: Direct stream APIs to the remote (primary) JobServiceGateway.
```
same as above

fabiokung commented on titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java:
```diff
@@ -1,64 +1,272 @@
+    private Id getMetricId(String metricName, Throwable t) {
+        String reason = t.getClass().getSimpleName();
+
+        if (t instanceof StatusRuntimeException) {
```
Remove this type check, `Status.fromThrowable()` already does it for you with the correct types. You might need to refactor this code a bit to recover the original message if `code == UNKNOWN`

fabiokung reviewed (approved):
LGTM with some minor nits

fabiokung commented:
> Are there other streaming APIs apart from `observeJob` and `observeJobs` which I should also avoid falling back on?

No, I believe these are the two gRPC streaming APIs we currently serve on the JobManagement service
